src/pdn/src/PdnGen.cc

```diff
@@ -225,14 +251,629 @@ PdnGen::addGlobalConnect(dbBox* region, const char* instPattern, const char* pin
+      for (auto& shape : shapes) {
+        all_shapes_layer.insert(shape);
+      }
+    }
+  }
+
+  if (trim) {
+    trimShapes();
+
+    cleanupVias();
+  }
+}
+
+void PdnGen::cleanupVias()
+{
+  debugPrint(logger_, utl::PDN, "Make", 2, "Cleanup vias - begin");
+  for (auto* grid : getGrids()) {
+    grid->removeInvalidVias();
+  }
+  debugPrint(logger_, utl::PDN, "Make", 2, "Cleanup vias - end");
+}
+
+void PdnGen::trimShapes()
+{
+  debugPrint(logger_, utl::PDN, "Make", 2, "Trim shapes - start");
+  auto grids = getGrids();
+
+  std::vector<ViaPtr> all_vias;
+  for (auto* grid : grids) {
+    grid->getVias(all_vias);
+  }
+
+  for (auto* grid : grids) {
+    for (const auto& [layer, shapes] : grid->getShapes()) {
+      for (const auto& [box, shape] : shapes) {
+        shape->clearVias();
+      }
+    }
+  }
+
+  for (const auto& via : all_vias) {
+    via->getLowerShape()->addVia(via);
+    via->getUpperShape()->addVia(via);
+  }
+
+  for (auto* grid : grids) {
+    if (grid->type() == Grid::Existing) {
+      // fixed shapes, so nothing to do
+      continue;
+    }
+    const auto& pin_layers = grid->getPinLayers();
+    for (const auto& [layer, shapes] : grid->getShapes()) {
+      for (const auto& [box, shape] : shapes) {
+        if (!shape->isModifiable()) {
+          continue;
+        }
+        // only trim shapes on non pin layers
+        if (pin_layers.find(shape->getLayer()) != pin_layers.end()) {
+          continue;
+        }
+
+        Shape* new_shape = nullptr;
+        const odb::Rect new_rect = shape->getMinimumRect();
+        if (new_rect == shape->getRect()) {  // no change to shape
+          continue;
+        }
+
+        // check if vias and shape form a stack without any other connections
+        bool effectively_vias_stack = true;
+        for (const auto& via : shape->getVias()) {
+          if (via->getArea() != new_rect) {
+            effectively_vias_stack = false;
+            break;
+          }
+        }
+        if (!effectively_vias_stack) {
+          new_shape = shape->copy();
+          new_shape->setRect(new_rect);
+        }
+
+        auto* component = shape->getGridComponent();
+        if (new_shape == nullptr) {
+          if (shape->isRemovable()) {
+            component->removeShape(shape.get());
+          }
+        } else {
+          component->replaceShape(shape.get(), {new_shape});
+        }
+      }
+    }
+  }
+  debugPrint(logger_, utl::PDN, "Make", 2, "Trim shapes - end");
+}
+
+VoltageDomain* PdnGen::getCoreDomain()
+{
+  if (core_domain_ == nullptr) {
+    setCoreDomain(nullptr, nullptr, {});
+  }
+
+  return core_domain_.get();
+}
+
+std::vector<VoltageDomain*> PdnGen::getDomains()
+{
+  std::vector<VoltageDomain*> domains;
+  domains.push_back(getCoreDomain());
+
+  for (const auto& domain : domains_) {
+    domains.push_back(domain.get());
+  }
+  return domains;
+}
+
+std::vector<VoltageDomain*> PdnGen::getConstDomains() const
```

Comment:
What is the difference from getDomains()?

Reply:
`getDomains` will make the initial domain if it's not been defined, `getConstDomains` will only return the domains that exist. I might be able to clean that up to make them a single function.